### PR TITLE
copr: implement collation aware Vec-SlowHashAggr

### PR DESCRIPTION
**Pull request by andylokandy**
###  What have you changed?

Based on https://github.com/tikv/tikv/pull/6608

Let Vec-SlowHashAggr executor regard the collation of ordering expressions.

###  What is the type of the changes?

- New feature (a change which adds functionality)

###  How is the PR tested?

- Unit test

### CODE REVIEW

**Comment by TennyZhuang** [comment]
The collation API change is awesome

**Comment by andylokandy** [comment]
@TennyZhuang @breeswish PTAL

File: components/tidb_query/src/codec/collation/utf8mb4.rs [hunk 1]
```diff
@@ -400,24 +404,27 @@ impl Collator for CollatorUtf8Mb4Bin {
...
+    fn write_sort_key<W: BufferWriter>(bstr: &[u8], writer: &mut W) -> Result<usize> {
+        str::from_utf8(bstr)?;
         writer.write_bytes(bstr)?;
         Ok(bstr.len())
     }
 
     #[inline]
-    fn sort_compare(a: &[u8], b: &[u8]) -> Result<Ordering, DecodeError> {
+    fn validate(bstr: &[u8]) -> Result<()> {
+        str::from_utf8(bstr)?;
```
Comment (breeswish) [inline components/tidb_query/src/codec/collation/utf8mb4.rs]: We already have validate() function, why do we need to validate again in `write_sort_key` and `sort_compare`?

File: components/tidb_query/src/codec/data_type/vector.rs [hunk 2]
```diff
@@ -351,6 +351,38 @@ impl VectorValue {
             }
         }
     }
+
+    pub fn encode_sort_key(
```
Comment (breeswish) [inline components/tidb_query/src/codec/data_type/vector.rs]: The name might need to to be changed. this one not only encode the sort key, but only encode other kind of datums.

Comment (andylokandy) [reply components/tidb_query/src/codec/data_type/vector.rs]: I think all kinds of data can be considered as sort key, but `Bytes`, as a special case, is different from the original data.

File: components/tidb_query/src/codec/collation/utf8mb4.rs (hunk 1 above)
Comment (andylokandy) [reply components/tidb_query/src/codec/collation/utf8mb4.rs]: In fact, validate is not yet used... and I may remove it in later pr, if it's useless in fashhashaggr.

Comment (bb7133) [reply components/tidb_query/src/codec/collation/utf8mb4.rs]: Do we have similar validation for UTF8 strings before(I mean, before collations)?

**Pull request merged**